### DISCUSSION TITLE
Can't access tools and live audio from my Birdnet-pi domain name

### DISCUSSION THREAD

**Discussion by blogdemoi** [mcguirepr89/BirdNET-Pi#508] [Q&A]
Hi,

I can access the tools and live audio on my local network, but can't on my domain name. It asks for a username (pi) and password in a pop up small window but when I enter them, nothing happens and the pop up window reloads...

What could be wrong?

Thank you.

**Comment by mcguirepr89** [comment]
Hi, @blogdemoi -- if nothing happens and the popup window reloads, it sounds like the wrong credentials.

At some point did you manually change the authentication username from [its default user `birdnet`](https://github.com/mcguirepr89/BirdNET-Pi#access) to the `pi` user as mentioned?
>It asks for a username (pi) and password in a pop up small window but when I enter them, nothing happens and the pop up window reloads...

**Reply by blogdemoi** [reply]
No, I didn't change anything. I just use the user pi to ssh in Birdnet-Pi.

**Reply by mcguirepr89** [reply]
Then it sounds like we'll need to troubleshoot this with some logging.

If you could please do the following, it will tell us exactly why you are unable to login from your domain.

1. SSH into the BirdNET-Pi
2. Issue:
   ```
   journalctl -fu caddy
   ```
3. Attempt to access "Tools" or the "Live Audio" from your domain.
4. Paste the output from your `caddy` log here for us to review.

My very best,
Patrick

**Comment by blogdemoi** [comment]
> Paste the output from your caddy log here for us to review.

Could you tell me where to find the logs?

**Reply by mcguirepr89** [reply]
Step 2 will provide log output

**Comment by blogdemoi** [comment]
```
birdnet:~/BirdNET-Pi$ journalctl -fu caddy
-- Journal begins at Fri 2022-08-19 17:17:01 CEST. --
Aug 22 17:31:37 birdnet caddy[794]: {"level":"info","ts":1661182297.452636,"msg":"serving initial configuration"}
Aug 22 17:31:37 birdnet caddy[794]: {"level":"info","ts":1661182297.4552784,"logger":"tls.cache.maintenance","msg":"stopped background certificate maintenance","cache":"0x40004e6000"}
Aug 22 17:31:37 birdnet caddy[794]: {"level":"info","ts":1661182297.9329407,"logger":"admin","msg":"stopped previous server","address":"tcp/localhost:2019"}
Aug 22 17:31:37 birdnet caddy[794]: {"level":"info","ts":1661182297.9339142,"msg":"shutdown complete","signal":"SIGTERM","exit_code":0}
Aug 22 17:31:37 birdnet systemd[1]: caddy.service: Failed with result 'timeout'.
Aug 22 17:31:37 birdnet systemd[1]: Failed to start Caddy.
Aug 22 17:31:37 birdnet systemd[1]: caddy.service: Consumed 1min 20.184s CPU time.
-- Boot 45578b4401ed448598c917eb675d5a64 --
Aug 22 18:30:31 birdnet caddy[768]: {"level":"error","ts":1661185831.7942605,"logger":"http.handlers.reverse_proxy","msg":"aborting with incomplete response","error":"context canceled"}
Aug 22 18:38:31 birdnet caddy[768]: {"level":"error","ts":1661186311.871622,"logger":"http.handlers.reverse_proxy","msg":"aborting with incomplete response","error":"context canceled"}
Aug 22 19:14:50 birdnet caddy[768]: {"level":"error","ts":1661188490.1509645,"logger":"http.log.error","msg":"EOF","request":{"remote_ip":"10.0.242.137","remote_port":"44788","proto":"HTTP/1.1","method":"GET","host":"birdnet.local","uri":"/scripts/config.php?restart_php=true","headers":{"Authorization":[],"Connection":["keep-alive"],"Referer":["http://birdnet.local/views.php?latitude=48.8422&longitude=2.1823&birdweather_id=&apprise_input=mailtos%3A%2F%2Farnauld%3AkO6fvqXWkd95G6BhrYFpEBHjIiVkzV%40arnauld.org%3A587%3Fsmtp%3Darnauld.org%0D%0A&apprise_notification_title=New+BirdNET-Pi+Detection&apprise_notification_body=A+%24sciname+%24comname+was+just+detected+with+a+confidence+of+%24confidence&apprise_notify_new_species=on&apprise_weekly_report=on&flickr_api_key=bf73bf58c4d5b0bfa056f89644a728ac&flickr_filter_email=&language=fr&timezone=Europe%2FParis&status=success&submit=settings&view=Settings"],"User-Agent":["Mozilla/5.0 (Windows NT 10.0; rv:103.0) Gecko/20100101 Firefox/103.0"],"Accept":["*/*"],"Accept-Language":["en-US,en;q=0.5"],"Dnt":["1"],"Accept-Encoding":["gzip, deflate"],"Cookie":[],"Sec-Gpc":["1"]}},"duration":9.239384854,"status":502,"err_id":"bxr3punie","err_trace":"reverseproxy.statusError (reverseproxy.go:1184)"}
```

**Reply by mcguirepr89** [reply]
Looks like Caddy is not starting at all, (or that is an incomplete log) -- I also don't see the basic auth login attempt from a domain (though I do see something from a public IP address `"request":{"remote_ip":"10.0.242.137","remote_port":"44788",` -- not sure whether that is related to reaching the installation via domain, and I can't say why it is reporting at all on port 44788).

Please perform the following steps to ensure we are getting the full log:

1. SSH into the BirdNET-Pi
2. Issue `sudo reboot`
3. When the BirdNET-Pi boots back up, try accessing the live audio stream and "Tools" page as you mention not working
4. Finally, issue `journalctl --no-pager --boot -ucaddy`

Those steps will:
1. Give you access to the device
2. Clear out the journal
3. Populate the journal with a failed attempt to access the "Tools" and "Live Audio"
4. Print out the journal's content ONLY showing what has happened since the most recent boot

**Reply by blogdemoi** [reply]
Hi,

Here are the results:
```

birdnet:~/BirdNET-Pi$ journalctl --no-pager --boot -ucaddy
-- Journal begins at Sat 2022-08-20 18:17:02 CEST, ends at Tue 2022-08-23 23:08:41 CEST. --
Aug 23 23:04:56 birdnet systemd[1]: Starting Caddy...
Aug 23 23:05:00 birdnet caddy[750]: caddy.HomeDir=/var/lib/caddy
Aug 23 23:05:00 birdnet caddy[750]: caddy.AppDataDir=/var/lib/caddy/.local/share/caddy
Aug 23 23:05:00 birdnet caddy[750]: caddy.AppConfigDir=/var/lib/caddy/.config/caddy
Aug 23 23:05:00 birdnet caddy[750]: caddy.ConfigAutosavePath=/var/lib/caddy/.config/caddy/autosave.json
Aug 23 23:05:00 birdnet caddy[750]: caddy.Version=v2.5.2 h1:eCJdLyEyAGzuQTa5Mh3gETnYWDClo1LjtQm2q9RNZrs=
Aug 23 23:05:00 birdnet caddy[750]: runtime.GOOS=linux
Aug 23 23:05:00 birdnet caddy[750]: runtime.GOARCH=arm64
Aug 23 23:05:00 birdnet caddy[750]: runtime.Compiler=gc
Aug 23 23:05:00 birdnet caddy[750]: runtime.NumCPU=4
Aug 23 23:05:00 birdnet caddy[750]: runtime.GOMAXPROCS=4
Aug 23 23:05:00 birdnet caddy[750]: runtime.Version=go1.18.3
Aug 23 23:05:00 birdnet caddy[750]: os.Getwd=/
Aug 23 23:05:00 birdnet caddy[750]: LANG=en_GB.UTF-8
Aug 23 23:05:00 birdnet caddy[750]: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Aug 23 23:05:00 birdnet caddy[750]: NOTIFY_SOCKET=/run/systemd/notify
Aug 23 23:05:00 birdnet caddy[750]: HOME=/var/lib/caddy
Aug 23 23:05:00 birdnet caddy[750]: LOGNAME=caddy
Aug 23 23:05:00 birdnet caddy[750]: USER=caddy
Aug 23 23:05:00 birdnet caddy[750]: INVOCATION_ID=cf52ce1f6e594892bac3525bc20410fc
Aug 23 23:05:00 birdnet caddy[750]: JOURNAL_STREAM=8:13904
Aug 23 23:05:00 birdnet caddy[750]: {"level":"info","ts":1661288700.0989685,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
Aug 23 23:05:00 birdnet caddy[750]: {"level":"info","ts":1661288700.2022178,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["//[::1]:2019","//127.0.0.1:2019","//localhost:2019"]}
Aug 23 23:05:00 birdnet caddy[750]: {"level":"info","ts":1661288700.2053537,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x400018a9a0"}
Aug 23 23:05:00 birdnet caddy[750]: {"level":"warn","ts":1661288700.207511,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}
Aug 23 23:06:18 birdnet caddy[750]: {"level":"info","ts":1661288778.0611763,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/var/lib/caddy/.local/share/caddy"}
Aug 23 23:06:18 birdnet caddy[750]: {"level":"info","ts":1661288778.080152,"logger":"tls","msg":"finished cleaning storage units"}
Aug 23 23:06:18 birdnet caddy[750]: {"level":"info","ts":1661288778.0823724,"msg":"autosaved config (load with --resume flag)","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Aug 23 23:06:18 birdnet caddy[750]: {"level":"info","ts":1661288778.1042163,"msg":"serving initial configuration"}
Aug 23 23:06:18 birdnet systemd[1]: Started Caddy.
birdnet:~/BirdNET-Pi$
```

**Reply by mcguirepr89** [reply]
Okay, let me just make sure I'm understanding things clearly:

You _are_ able to access the Live Audio and Tools from http://birdnetpi.local

but you _are not_ able to access the Live Audio and Tools from your domain?

Is that correct?

Is there a reverse_proxy between your installation and the internet?

**Comment by blogdemoi** [comment]
> You are able to access the Live Audio and Tools from http://birdnetpi.local/

yes

> but you are not able to access the Live Audio and Tools from your domain?

yes

> Is that correct?

yes

> Is there a reverse_proxy between your installation and the internet?

Yes, there is a reverse proxy. What I can tell you is that I have first my Internex Box (provided by my ISP). Attached to this box is a first Raspberry Pi running a web server (where I installed and run nextcloud, a blog, my e-mails, etc.).

When I decided to install Birdnet-Pi on another Raspberry-Pi I couldn't have it exposed to the outside world because my Internet Box only accepts one web server behind it.

Fortunately, the first Raspberry Pi running my home server runs Yunohost https://yunohost.org/en/whatsyunohost (an easy install server solution) which has an application called Redirect https://github.com/YunoHost-Apps/redirect_ynh that allows me to redirect Birdnet-Pi and this way I can have two web servers exposed to the outside world.

**Reply by mcguirepr89** [reply]
>Yes, there is a reverse proxy.

Then this is the answer -- headers for the basic auth are not being forwarded by your reverse proxy properly. You'll need to make sure the headers for the basic auth are going to your domain with the proper protocol and matching your domain name, otherwise Caddy will reject them. 

I'm sorry I don't have a quick and easy way to troubleshoot this -- you'll need to use WireShark or run some packet captures or something along those lines to get a sense of what the basic auth headers look like when they leave your reverse proxy and be sure they're getting forwarded properly.

**Reply by blogdemoi** [reply]
That's ok. I thank you very much to take time trying to help me. :-) I will seek help on the Yunohost forums and anyway I don't really need live audio or accessing tools when I am not at home. My Birdnet-Pi has been running for 4 days already and I am enjoying it very much, it is so interesting.

Best Regards,

**Reply by mcguirepr89** [reply]
You're very welcome -- one option that I can endorse (among the many reasons I've chosen `caddy` for the webserver) is `caddy`'s production-ready `reverse_proxy` ability.

If you were to refactor your other webserver so that it were serving only on `localhost:9090`, you could easily have a ready-to-go reverse proxy [installing caddy](https://caddyserver.com/docs/install#debian-ubuntu-raspbian), and then using a Caddyfile like the one below:

```
mydomain.com {
    reverse_proxy * localhost:9090
}

mybirdnetpi.com {
    reverse_proxy * birdnetpi.local:80
}
```

(ensuring that the `/etc/hosts` file has an entry for `birdnetpi.local` just to be safe)

Caddy would automatically fetch SSL certificates for both websites, and will force redirect HTTP traffic to be HTTPS -- it would also renew the certificates before they expire.

The [Caddy documentation](https://caddyserver.com/docs/) is a plentiful in case you decide to go that route.

My very best regards,
Patrick
